### ISSUE TITLE
Proposal: Parametric sagas

### ISSUE THREAD

**Issue by aikoven** [yelouafi/redux-saga#43]
I've run into following problem:
Suppose we have two sagas: one daemon, and one `call`able:
```js
function *daemonSaga() {
  while (true) {
    const action = take(...);

    yield call(callableSaga, action.payload);
  }
}

function *callableSaga(params) {
  // ...
}
```

Now, suppose that `callableSaga` becomes dependent on state (in my particular case it depends on locale setting that is stored in the state). We could use `getState()` in `daemonSaga` and pass extra state  data to `callableSaga`, but that's not very convenient: if there is another dependency in `callableSaga`, we'd have to change code of `daemonSaga` again; also, if there is another saga that `call`s `callableSaga`, we'd have to duplicate code for extracting state data. It seems that it has to be `callableSaga`'s responsibility to get its state dependencies.

So my first approach was to just pass `getState` to `callableSaga`:
```js
function *daemonSaga(getState) {
  // ...
    yield call(callableSaga, action.payload, getState);
  // ...
}

function *callableSaga(params, getState) {
  // ...
}
```

Although it still has some drawbacks: every saga that `call`s `callableSaga` must accept `getState` argument even if it doesn't have dependencies on state itself; if nested `call` level is longer then we have to pass `getState` from the very top.

Sounds familiar, isn't it? (Hint: [react-redux](https://github.com/rackt/react-redux)).

There's another concern about  `getState` argument for sagas: they become dependent on state shape which breaks encapsulation.

## Proposal
* Allow sagas to be `connect`ed to the store, similar to how React components are connected by react-redux. If [function decorators](https://github.com/wycats/javascript-decorators/issues/4) were supported, it could look like this:
```js
@connectSaga(state => /* extract params */)
function* connectedSaga(getParams) {
  // ...
}
```
Decorator returns regular saga with `getState` argument. Added benefit here is that we can decouple saga from state shape by using selectors.
* Allow sagas to call connected sagas without having to pass `getState`:
```js
function *saga() {
  yield run(connectedSaga);
}
```
* Allow connected sagas to accept extra parameters. They can be handled in similar way as react-redux handles component props: merge them with result of selector:
```js
@connectSaga(state => ({stateParam: ...}))
function* connectedSaga(getParams) {
  const {stateParam, extraParam} = getParams();
  // ...
}

function *saga() {
  yield run(connectedSaga, {extraParam: ...});
}
```

**Comment by aft-luke** [comment]
I too struggle with getState in any sort of action creator or side effect, whether that be redux-thunk or redux-saga... it just smells wrong that the details of the state shape have to leak into a component that just shouldn't care.

This is interesting, selecting a state slice from the saga. It's a tradeoff of more boilerplate for a centralized location where we can change state, but is that tradeoff worth it? I'm not sure.

One approach I've tried is normalizing my state for components where I have not been able to avoid `getState()` inside a side effect. The normalization puts all these state objects inside a `state._entities` and then offers up the id when dispatching. This, of course, requires assigning targets to events (which is a prerequisite of reusable components across the app anyways). so there's a bit more complexity here required. All in all it offers a nice API, though:

```
// thunk example
actionCreator() {
  return (dispatch, getState, target) {
    let relevantStateSlice = getState()._entities.get(target.id)
    // do stuff with the state slice ...
  }
}

**Comment by aikoven** [comment]
In my case I first fetch some entities and then parse them according to locale setting. Entities and settings are stored in different branches of state tree, so I can't do the parsing inside `entities` reducer as it can't see anything outside of `entities` branch. I also don't like the idea to have a global reducer that can see the whole state.

The only place that seems right for it is saga (or thunk).

But there is a problem with `getState`: sagas (and thunks) must be aware of state shape. I was thinking about not using `getState` at all, but then I have no place where I can do my data processing that depends on different state branches.

We already have nice and composable way to collect dependencies from state — [selectors](https://github.com/rackt/reselect). We can also reuse selectors between connected components and connected sagas.

This way the only place that depends on state structure is selectors; everything else uses selectors as a unique interface to the state. And reducers remain locked to their own branch.

I've already implemented custom middleware that allows this approach and now I'm refactoring existing sagas. There isn't much boilerplate because I mostly reuse existing selectors.

**Comment by yelouafi** [comment]
@aikoven Can't you pass the required data (entities & locale setting) in the action payload ? Then the only dependency you'll have to the 2 selectors will be in the component firing the action.

**Comment by aikoven** [comment]
@yelouafi I did that before, but it doesn't feel right that component has to query some state data that's not needed by component itself, only to pass it to saga via action. Again, if that saga would eventually require another dependency, we'd have to update every component that fires that action.

**Comment by yelouafi** [comment]
@aikoven Sorry for the late response. I was hopping for some more comments from other people

I thinks the issue you raised is valid. I also remember this was also raised by some Redux users. I agree the idea of a master reducer feels like a code smell. Although i"m not sure it belongs to the Saga to this kind of things. The point is that Sagas handle effects, and effects are about yielding something at some specific moment. Processing response is normally done inside api services which have to return a final value to the Saga. 

But in the same time I can't think as you said of another place where to put this. So the question question I'm asking (myself principally) is: is it right to implement this *by default inside Sagas* ?  Or maybe the Saga acts in this situation like a *bridge* between the store and api services. i.e. gets the required info from the store and pass it to api services as arguments.

My other concern is regarding the implementation Can we implement this without having a dependency to a Redux concept (getState) inside the internal proc method ? (which right now is independent). Keeping proc independant from Redux allows it to be used in other situations ( like connecting runSaga to external IO ).

Another concern is related to effect logging, if we wrap the Saga inside a decorator, do we loose the name of the generator function ? Because if so, the effect monitoring wont be able to print a readable trace of the Sagas (I'm not well versed into decorators yet)

**Comment by aikoven** [comment]
@yelouafi My current view is aligned with what you wrote: Saga is responsible for extracting all data required for API call. 

There is ambiguity in whether it should process response or processing should be done in API service, but I think that it depends on how complex your service is. If every call is just simple `fetch()`, then maybe it doesn't make much sense to write a service to wrap it. But when it becomes more complicated, probably it's best to move that logic out of Saga.

Any way if response processing requires some extra data from the store then we have to do extraction inside Saga.

Another ambiguity is separation of concerns between Saga and reducers: nothing stops us from calculating next state inside Saga and just passing it to reducer via action. On the other hand we could put response plus extra state data into action and do all processing in reducer. In my preferred model **reducers are responsible only for create/update/delete in their specific state tree branches**, without any other logic, and **action payload contains minimal amount of data needed for reducer**.

Considering implementation: there's no need to couple `proc()` with Redux. We can use `dispatch` argument of `proc` to pass child saga up to the middleware:
```js
const CONNECTED_SAGA = Symbol();

function run(connectedSaga, params) {
  // replace with new effect type:
  return put({
    [CONNECTED_SAGA]: true,
    connectedSaga,
    params,
  });
}

function sagaMiddleware(...sagas) {
  return ({getState, dispatch}) => {
    const sagaEmitter = emitter();

    const wrappedDispatch = (action) => {
      if (action[CONNECTED_SAGA]) {
        return proc(
          action.connectedSaga(getState, action.params),
          sagaEmitter.subscribe,
          wrappedDispatch
        ).done;
      } else {
        return dispatch(action)
      }
    };

    sagas.forEach(saga=> {
      proc(
        saga(getState),
        sagaEmitter.subscribe,
        wrappedDispatch
      )
    });

    return next => action => {
      const result = next(action);
      sagaEmitter.emit(action);
      return result;
    }
  };
}
```

Considering decorators: from the discussion I linked to it seems that function decorators won't be supported any time soon. So maybe `connectSaga` should return some wrapper object, like 
```js
{
  selector: function (getState) {...}, 
  saga: function generatorName(getParams) {...}
}
```
Or it can assign `selector` property to the generator function.

**Comment by aikoven** [comment]
Just found out that you can actually change function name by 
```js
Object.assign(fn, 'name', {value: 'new name'})
```

**Comment by slorber** [comment]
@aikoven I think in all cases redux-saga does not have to be dependent on your redux state at all. However I discussed that with @yelouafi and we agree that it however can be more convenient.

Related discussion: https://github.com/yelouafi/redux-saga/issues/8

Also check this blogpost: http://jaysoo.ca/2016/01/03/managing-processes-in-redux-using-sagas/
Where I suggest the author to change the saga from:

```javascript
function* runTimer(getState) {
  // The sagasMiddleware will start running this generator.

  // Wake up when user starts timer.
  while(yield take('START')) {
    while(true) {
      // This side effect is not run yet, so it can be treated
      // as data, making it easier to test if needed.
      yield call(wait, ONE_SECOND);

      // Check if the timer is still running.
      // If so, then dispatch a TICK.
      if (getState().status === 'Running') {
        yield put(actions.tick());
      // Otherwise, go idle until user starts the timer again.
      } else {
        break;
      }
    }
  }
}
```
to

```javascript
function* runTimer() {
  while(yield take('START')) {
    while(true) {
      const {stop, tick} = yield race({
        stop   : take('STOP'),
        tick : call(wait, ONE_SECOND);
      })
      if ( !stop ) {
        yield put(actions.tick());
      } else {
        break;
      }
    }
  }
}
```

You can see that finally no `getState()` is required at all.

The problem is that you consider that Redux is the source of truth, instead of considering the event log. There is no need to depend on the Redux store state because that state is already a projection of the event log. Whoever has access to that log is actually capable of rebuilding Redux store. It is the only dependency required, because all data you need is already is that log.

However if you have a reducer to compute some value over time in a Redux store, I understand this seems quite unnatural to  also compute that same value as a local variable inside the saga: 
- It duplicates the data (inside Redux store + inside saga as local variable)
- It duplicates the code to compute that value (generators are not really written like reducers...)


Is data duplication really a problem? Sagas come from the backend world, CQRS and EventSourcing. In this world, the saga and reducers you are familiar with are actually often running on different physical machines, and it is not effective at all to query each others: doing `getState()` or querying selector data into saga involves IO operations and make the saga unefficient. The only thing these physical machines share is the event log, and they don't query each others. It just works in an efficient and a very decoupled way as the only real dependency is the event log. Each physical machine host a microservice and implements a reducer (but actually persisted in a local DB/index) and some selectors (queries over the DB/index). In this world the saga can run alone in a single physical machine, and has no access to any microservice DB / query API. The saga can itself has its own local DB.

I don't think data duplication is a problem. A reducer is very predictible and given a state + an event it's supposed to always return the same result, so using the same reducer multiple times inside a single app is not a big deal. However unlike the backend world we are running on a single browser so it should be performant enough... But the reducer could be memoized too (like reselect we could create rereduce :p) 


**Final thoughts**

@aikoven I think in your example with entities/settings, there are different solution. One is to consider settingsReducer as a shared/common dependency of your app (because you need these settings in many places) and then you could use settingsReducer  inside entitiesReducer even if it creates duplication. This would permit you to parse the entity in an atomic/sync way while using a saga for this means the entity would be first unparsed and then become parsed.

Another solution, using a saga, that seems more appropriate to me is that instead of trying to inject Redux's store state inside the saga, why not letting the saga declare that it actually needs data that is computed by a reducer (ie registering reducer dependencies to a saga). I like the idea that the saga remains free of querying state from the Redux store, and that the saga should be able to compute its state on its own from the event log. However it would be convenient if the saga could reuse code that is already implemented for redux stores to project the event log. I prefer to depend on pure shared reducers than state shape / selectors.

Instead of:

 ```javascript
@connectSaga(state => ({stateParam: ...}))
function* connectedSaga(getParams) {
  const {stateParam, extraParam} = getParams();
  // ...
}

function *saga() {
  yield run(connectedSaga, {extraParam: ...});
}
```

It could look like

 ```javascript
@connectSaga({ stateParam: stateParamReducer })
function* connectedSaga(params) {
  const {stateParam, extraParam} = params;
  // ...
}

function *saga() {
  yield run(connectedSaga, {extraParam: ...});
}
```

**Comment by aikoven** [comment]
@slorber Thanks for elaborative answer. I get your point.

I agree that there are some drawbacks with this approach. One more that I already encountered is that state may change between effects which makes Sagas harder to test, i.e.:
```js
function* saga(getParams) {
  yield ...
  const params1 = getParams();
  yield ...
  const params2 = getParams();  // may differ from params1
}
```

I'd like to have Sagas completely decoupled from state, still there are several cases that need to be resolved:
* Replaying event log only works for sagas that were started on app startup. Which means that subroutines can't be connected. They can probably be refactored, something like:
```js
// before
function* subroutine(getParams) {
  return yield call(service, getParams().param)
}

function* saga() {
  while(true) {
    yield run(subroutine);
  }
}

// after
function* subroutineSaga(getParams) {
  while(true) {
    yield take(START_SUBROUTINE);
    const payload = yield call(service, getParams().param);
    yield put({type: SUBROUTINE_DONE, payload})
  }
}

function* saga() {
  while(true) {
    yield put({type: START_SUBROUTINE});
    const result = (yield take(SUBROUTINE_DONE)).payload;
  }
}
```
Although this is very verbose (which may be ok considering gained decoupling), but more importantly, there

**Comment by aikoven** [comment]
@slorber Thanks for elaborative answer. I get your point.

I agree that there are some drawbacks with this approach. One more that I already encountered is that state may change between effects which makes Sagas harder to test, i.e.:
```js
function* saga(getParams) {
  yield ...
  const params1 = getParams();
  yield ...
  const params2 = getParams();  // may differ from params1
}
```

I'd like to have Sagas completely decoupled from state, still there are several cases that need to be resolved:
* Replaying event log only works for sagas that were started on app startup. Which means that subroutines can't be connected. They can probably be refactored, something like:
```js
// before
function* subroutine(getParams) {
  return yield call(service, getParams().param)
}

function* saga() {
  while(true) {
    const result = yield run(subroutine);
  }
}

// after
function* subroutineSaga(getParams) {
  while(true) {
    yield take(START_SUBROUTINE);
    const payload = yield call(service, getParams().param);
    yield put({type: SUBROUTINE_DONE, payload})
  }
}

function* saga() {
  while(true) {
    yield put({type: START_SUBROUTINE});
    const result = (yield take(SUBROUTINE_DONE)).payload;
  }
}
```
Although this is very verbose (which may be ok considering gained decoupling), but more importantly, there is no way to run several instances of subroutine in parallel.

A workaround for this would be to just pass subroutine dependencies via arguments, but it would become a mess if parent saga calls many subroutines each having dependencies.

Another solution I see is to be able to clone running saga with all its internal state (similar to Unix `fork`). This way parent saga can clone subroutine before `put`ting `START_SUBROUTINE`. But again we'd introduce coupling between parent saga and subroutine.

Third solution is to extract another routine from `subroutineSaga`:
```js
function* subsubroutine(param) {
  const payload = yield call(service, getParams().param);
  yield put({type: SUBROUTINE_DONE, payload})
}

function* subroutineSaga(getParams) {
  while(true) {
    yield take(START_SUBROUTINE);
    yield fork(subsubroutine, getParams().param);
  }
}
```
Which makes it even more verbose 😫.

* If we connect Saga using reducer, we'd again rely on state shape. Not global state, only one branch, but still. But here we can use selectors that accept this substate.

I'll try to refactor my code with new approach to see if I can come up with more ideas.

Still I'm not sure what are real benefits in having slice of state recomputed locally for each saga. In any way we'd have to query the state, even if it is local.

**Comment by slorber** [comment]
@aikoven I'm just trying to think about how backend systems implement saga. I'm not experienced enough in building those kind of components but as far as I know there's not really a concept of "subroutine". It looks like your subroutine is a way to factorise code and split a complex saga into smaller bits, bit does it mean that there are multiple sagas instead of a single big one then? Like React we have smart/dumb components. Wouldn't it make sens to only manage state in a "smart generator" (ie the saga) and the factorized/reusable subroutines could just receive required state as a parameter?

> Still I'm not sure what are real benefits in having slice of state recomputed locally for each saga. In any way we'd have to query the state, even if it is local.

If you consider the saga runs on a separate computer and can't query other computers, the only solution to get this required state is to compute it locally from the event log. A reducer is not a "slice of state" it's just pure code to project an event log into a meaningful state snapshot. That reducer does not necessarily have to work from the very beginning of the event log. 

It is ok for the saga to query its local state as it manages its own state. In backends sagas can be backed by databases and they can query these databases as long as the saga itself has the responsability to manage that database. I mean that it's the saga that creates and manage its own DB and can query it (because every state can't always fit in memory like we are used to on frontends). That DB is the saga DB only, and is meant to be queried only by the saga and no one else.

What I want to say is that each component (including sagas) that has acces to the event log picks the state it's interested in from it, by using code that can take form of a reducer, or a generator, or whatever fancy stuff you want. If 2 components of the system need to pick the same data from the log, doesn't it make sens to factorise the code to do so? Once they picked it, they manage that state the way they want and store is where they want to in an autonomous way. These components can be implemented in different languages and technologies, using actors, state machines, sagas, generators, whatever... If one component gets totally refactored, or even deleted, the others are not affected at all (unless the event log is affected). A team can take ownership of a component without having to understand how other components are working nor their component query API updates and lifecycle management. This is not the case if you allow components to query each others.



So if we agree components should not query each others, the problem is that components can project the log to a state snapshot in different ways (ie reducers / generators). If you want to be able to reuse some of that logic it should share the same technology. It seems to me there would be less friction introducing reducers in sagas code than introducing generators in redux.

I don't really have an answer to these questions :)

**Comment by aikoven** [comment]
> Like React we have smart/dumb components. Wouldn't it make sens to only manage state in a "smart generator" (ie the saga) and the factorized/reusable subroutines could just receive required state as a parameter?

This proposal enables nesting "smart" sagas inside "smart" sagas, just as you can do with react-redux. But I admit that there aren't many cases for deeply nested saga calls compared to react components.

> A reducer is not a "slice of state" it's just pure code to project an event log into a meaningful state snapshot.

What I meant by "slice" is the part of state that is computed by given reducer. So if I have `settingsReducer` that manages state of shape `{locale: ..., /* other data */}`, and connect it to Saga, then querying `state.locale` means relying on the shape of that part.

This is probably not an issue in this particular case, because `locale` is computed independently from other settings, so I can introduce `localeReducer` that only manages `locale` value.

Also I'm starting to think that exposing state shape to Saga is ok in case of local state, because it's already structured for that particular Saga means.

Here's my current vision on this:

We allow to connect Saga to arbitrary reducer. When Saga starts, we start background process that listens to the event log and computes local state. This state may be queried by `getState` argument, or by some new effect.

We can reuse reducers that manage parts of the state that we are interested in and combine them using `combineReducers`.

```js
@connectSaga(combineReducers({
  locale: localeReducer,
  other: ...
}))
function* saga() {
  // ...
  const {locale, other} = yield getState();
}
```

Also now I see that we can easily have connected sagas that aren't started on app startup, by supplying their reducers with current state as initial.

**Comment by slorber** [comment]
> What I meant by "slice" is the part of state that is computed by given reducer. So if I have settingsReducer that manages state of shape {locale: ..., /* other data */}, and connect it to Saga, then querying state.locale means relying on the shape of that part.

> This is probably not an issue in this particular case, because locale is computed independently from other settings, so I can introduce localeReducer that only manages locale value.

> Also I'm starting to think that exposing state shape to Saga is ok in case of local state, because it's already structured for that particular Saga means.

Exactly. A reducer is a function that returns something, so obviously if you decide to use that function, you should be aware of the kind of shape it returns. Obviously it's better to use the smallest possible function that gets the job done so if you only need to pick the locale you don't want to pass your saga your redux's root reducer and have your saga broken everytime the path the locale setting is refactored :) 

Be aware that not all code reuse / factorisation is always a good idea:
http://udidahan.com/2009/06/07/the-fallacy-of-reuse/
http://us3.campaign-archive2.com/?u=1090565ccff48ac602d0a84b4&id=92902a19e4&e=7f00539c7d

I would only connect to sagas very simple reducers that would preferably compute very simple reducers, mayby only returning primitive values or very simple objects.

**Comment by aikoven** [comment]
So here's what I've come up with:

```js
function connectReducer(reducer, saga) {
  const store = createStore(reducer);

  const storeSaga = function* storeSaga() {
    try {
      while (true) {
        const action = yield take('*');
        store.dispatch(action);
      }
    } catch (e) {
      if (e instanceof SagaCancellationException)
        return;

      throw e;
    }
  };

  const connectedSaga = function*() {
    const task = yield fork(storeSaga);

    try {
      yield call(saga, store.getState);
    } finally {
      yield cancel(task);
    }
  };

  Object.defineProperty(connectedSaga, 'name', {value: saga.name});

  return connectedSaga;
}
```

Each connected saga gets its own store which receives every action dispatched to the main Redux store. That local store state getter is passed to saga as `getState` argument. Wrapper `connectedSaga` has no arguments — it doesn't depend on main Redux state anymore.

I've split out `localeReducer` from `settingsReducer` and reused it for Saga:

```js
const fetchEntitiesSaga = connectSaga(localeReducer, 
  function* fetchEntitiesSaga(getState) {
    // ...
    const entities = yield call(service);
    const locale = getState();
    const parsedEntities = parse(entities, locale);
    // ...
  }
);
```

**Comment by slorber** [comment]
that seems nice to me :)

What happens if you use `connectSaga` multiple times last one wins right?

**Comment by aikoven** [comment]
> What happens if you use connectSaga multiple times last one wins right?
No, the first one will already have zero arguments, so connecting second time won't change anything, except for creating another store and starting background saga for it.

Unfortunately I don't know how to connect sagas that aren't started on app startup. Rebuilding their local state would require having entire event log to the time they start.

**Comment by aikoven** [comment]
For now this has been very convenient way to pull a part of Redux state to Saga. It also made me refactor my reducers to split them into smaller ones that manage smallest possible state slice. Which is great even if I didn't use redux-saga, as it makes it much easier to reason about update logic and also to test it. And when some update logic requires data from other parts of the state, I treat it as a side-effect instead of introducing reducer that manages larger slice.

I'm thinking about publishing this as a separate repo.

I guess this issue can be closed.

**Comment by yelouafi** [comment]
@aikoven Sorry for not replying to this. I was following your discussion with @slorber and trying to abstract out something more generic for redux-saga

I think Decorators would be great as a mean for general configuration of Sagas (e.g. indicate that a saga depends on some actions so the middleware can buffer those actions). However it seems function Decorators wont be supported anysoon (currently at stage 0 and only for function expressions).

My feel is that connecting Sagas for what you want seems to be part of a much general concern of dependency injection.

For the state, there was a time I considered adding some effect `yield getState/setState` in order allow testing the Saga state as well but didn't explore further this way.

> It also made me refactor my reducers to split them into smaller ones that manage smallest possible state slice. Which is great even if I didn't use redux-saga, as it makes it much easier to reason about update logic and also to test it.

>And when some update logic requires data from other parts of the state, I treat it as a side-effect instead of introducing reducer that manages larger slice.

Did you see this proposal in Redux repo

https://github.com/rackt/redux/issues/1315

It proposes a solution to this issue based on Generators as well

```javascript
function* c(state = 0, action) {
  const aState = yield a
  const bState = yield b
  return state + Math.abs(aState - bState)
}
```

>I'm thinking about publishing this as a separate repo.

I guess this would make more sens actually. We can see later if we can factor out something to integrate into redux-saga

**Comment by aikoven** [comment]
@yelouafi Yes, I'm following that discussion. 

Still current approach seems more clean for me, it makes everything decoupled: reducers in the tree are completely independent from each other and from sagas. 

However it can be that my app is just not complex enough yet. 

Anyway, it is obvious that this approach won't fit everyone, so it should be kept separate.

**Comment by slorber** [comment]
Also @aikoven take a look at https://github.com/slorber/rereduce
You can use memoization to avoid doing twice the same computation in your solution.
Also not sure it's a big deal if you have reducers that don't compute on the whole log but a slice of it.

@yelouafi I don't like so much the solution of @gaearon here, mostly because as far as I remember, the c generator has to be on the same level as a and b reducers. It couples too much on the shape of the state, so introducing a new dependency could mean we have to refactor the state shape, and it could also lead to have to put a lot of reducers on the same level just for dependency injection in a complex app. If you start to shape your state because of technical constraints of a lib I think it's not a good thing at all.

If we really are going to follow the "query the redux store to get state from it" approach, I'd rather inject selectors to the saga. This makes the saga and the store shape keep being decoupled. The injected selectors could be binded with getState as parameter so that you don't have to pass getState method everywhere

**Comment by yelouafi** [comment]
@slorber 
> mostly because as far as I remember, the c generator has to be on the same level as a and b reducers.

Didn't saw that. Is this a technical constraint ? (I can't really see how this is a requirement for the above solution to work)

**Comment by slorber** [comment]
On the initial proposal:

> Just like with regular combineReducers(), this pattern can be applied recursively, although only reducers “on the same level” (inside the same combineReducers() call) would be allowed to yield each other.

**Comment by aikoven** [comment]
@slorber My proposed approach aims to decouple every piece of computation and any dependency is resolved by passing computed results via actions. Decoupling reducer means that getting any data that is external to it is a _side-effect_ and thus must be handled by saga. So I'm currently doing all computation that has dependencies inside sagas.

But with _rereduce_ we can go in another direction, i.e. do more pure computation in reducers and minimize code in sagas. It would also allow us to keep action payloads to minimum. Actually I think it's better, because amount of code handling side-effects is smaller.

Still there remain two different possible ways of decoupling saga from state shape: connecting saga to reducer or to selector.

**Comment by yelouafi** [comment]
@aikoven What do you think of this proposition

@gaearon first proposed to provide a `yield getState()` effect so we do not pass the getState param around.

@slorber proposed a more general `yield select(selector)` Effect.

**Comment by aikoven** [comment]
@yelouafi 👍

**Comment by yelouafi** [comment]
FYI [released in 0.9.0](https://github.com/yelouafi/redux-saga/releases/tag/v0.9.0)

**Comment by aikoven** [comment]
Good work on feature and docs!
